rust_icu_uloc/src/lib.rs

```diff
@@ -63,55 +63,178 @@ impl TryFrom<&ffi::CStr> for ULoc {
     }
 }
 
+/// Generates a method to wrap ICU4C `uloc` methods that require a resizable output string buffer.
+///
+/// The various `uloc` methods of this type have inconsistent signature patterns, with some putting
+/// all their input arguments _before_ the `buffer` and its `capacity`, and some splitting the input
+/// arguments.
+///
+/// Therefore, the macro supports input arguments in both positions.
+///
+/// For an invocation of the form
+/// ```
+/// buffered_string_method_with_retry!(
+///     my_method,
+///     BUFFER_CAPACITY,
+///     [before_arg_a: before_type_a, before_arg_b: before_type_b,],
+///     [after_arg_a: after_type_a, after_arg_b: after_type_b,]
+/// );   
+/// ```
+/// the generated method has a signature of the form
+/// ```
+/// fn my_method(
+///     uloc_method: unsafe extern "C" fn(
+///         before_type_a,
+///         before_type_b,
+///         *mut raw::c_char,
+///         i32,
+///         after_type_a,
+///         after_type_b,
+///         *mut UErrorCode,
+///     ) -> i32,
+///     before_arg_a: before_type_a,
+///     before_arg_b: before_type_b,
+///     after_arg_a: after_type_a,
+///     after_arg_b: after_type_b
+/// ) -> Result<String, common::Error> {}
+/// ```
+macro_rules! buffered_string_method_with_retry {
```

Comment:
I have a certain issue with macros, which is that they are difficult to maintain, and will likely remain so.  For this reason I'd like in general to use them sparingly.

Now I have to meta-compile the code in my head to understand what is going on in here.  And my editor is certainly going to fall apart trying to figure out whether there are errors in the macro source itself.

Consider the following alternatives:

1. If you must make macros, would it be possible to confine the extent of the macros only to the functions that need arg reordering, e.g. make macros that expand into closures with the correct arg binding, where the common skeleton remains the same?
2. Or make two non-macro instances of this code that do the respective correct thing for each ordering?  (Given that there is quite some variety this is probably not an option at all)

Now, I am curious to give it a test drive here, because we would benefit a lot from auto-generated binding code; so I'll approve as-is, to see where it gets us. 

But in the future please try to avoid macro use as much as possible.

Reply:
FYI, IntelliJ Rust correctly expands these macros.

Regarding **1**, so far, I haven't been able to think of a way to do this, given how many methods' signatures are variable. It might require something like [variadic generics](https://github.com/rust-lang/rfcs/issues/376). I'm open to suggestions.

As you say, **2** is impractical because we'd end up with several very long functions with almost identical code.